### PR TITLE
Delegate types in the framework. Fixes #239

### LINKED ISSUE #239: Delegates: article on Strongly Typed Delegates
This is a sub-task for #63.

Create the article that discusses strongly typed delegates (using Func, Action, and Predicate)

### PR DESCRIPTION
**Pull request by BillWagner**
This relies on PR # 246

### CODE REVIEW

File: docs/languages/csharp/delegates-strongly-typed.md [hunk 1]
```diff
@@ -1,11 +1,96 @@
...
+
 [Previous](delegate-class.md)
 
-This article will cover the classes in the .NET framework
-that support delegates, and how those map to the `delegate`
-keyword.
+In the previous article, you saw that you create specific delegate
+types using the `delegate` keyword. 
+
+Delegates are much more useful by embracing and enforcing type
```
Comment (ghogen) [inline docs/languages/csharp/delegates-strongly-typed.md]: Delegates are much more useful than what, function pointers?  Or do you mean that they are more useful than they would be without type safety?  I'd rephrase.  Maybe "Delegates are useful also because they enforce type safety..."

File: docs/languages/csharp/delegates-strongly-typed.md [hunk 2]
```diff
@@ -1,11 +1,96 @@
...
+
+```cs
+public delegate void Action();
+public delegate void Action<in T>(T arg);
+public delegate void Action<in T1, in T2>(T1 arg1, T2 arg2);
+// Other variations removed for brevity.
+```
+
+The `in` modifier on the generic type argument is covered in the article
+on covariance.
```
Comment (ghogen) [inline docs/languages/csharp/delegates-strongly-typed.md]: can we add a link here to the article (if it exists?)

File: docs/languages/csharp/delegates-strongly-typed.md [hunk 3]
```diff
@@ -1,11 +1,96 @@
...
+Use one of the `Action` types for any delegate type that has a void return type.
+
+The framework also includes several generic delegate types that you can use for
+delegate types that return values:
+
+```cs
+public delegate TResult Func<out TResult>();
+public delegate TResult Func<in T1, out TResult>(T1 arg);
+public delegate TResult Func<in T1, in T2, out TResult>(T1 arg1, T2 arg2);
+// Other varaiations removed for brevity
```
Comment (ghogen) [inline docs/languages/csharp/delegates-strongly-typed.md]: spelling: variations

File: docs/languages/csharp/delegates-strongly-typed.md [hunk 4]
```diff
@@ -1,11 +1,96 @@
...
+You may notice that for any `Predicate` type, a structurally equivalent `Func`
+type exists For example:
+
+```cs
+Func<string, bool> TestForString;
+Predicate<string> AnotherTestForString;
+```
+
+You might think these two types are equivalent. They are not.
+These two variables cannot be used interchangeable. A variable of one type cannot
```
Comment (ghogen) [inline docs/languages/csharp/delegates-strongly-typed.md]: interchangeably

File: docs/languages/csharp/delegates-strongly-typed.md (hunk 2 above)
Comment (BillWagner) [inline docs/languages/csharp/delegates-strongly-typed.md]: It's not written yet. It's issue #72 How do you want to handle that?

**Comment by BillWagner** [comment]
Updated (other than one open comment), rebased and squashed.

File: docs/languages/csharp/delegates-strongly-typed.md (hunk 2 above)
Comment (ghogen) [inline docs/languages/csharp/delegates-strongly-typed.md]: Let's merge this now and we'll add a note to issue #72 to remind us about adding this link.
